### ISSUE TITLE
Support OpenID Connect providers (like Google)

### ISSUE THREAD

**Issue by rysiekpl** [rancher/rancher#10053]
Rancher already supports GitHub as an auth provider. Supporting standard [OpenID Connect](https://openid.net/connect/) providers, like [Keycloak](http://www.keycloak.org/) or [GitLab](https://gitlab.com/help/integration/openid_connect_provider.md), would be a boon to many organizations, letting them integrate Rancher into their existing SSO infrastructure.

gz#13527

SURE-4586

**Comment by duffyjay** [comment]
+1

**Comment by dhrp** [comment]
+1

**Comment by RonWilliams** [comment]
+1 Also looking for this feature.

**Comment by L30Bola** [comment]
+1 this would be awesome.

**Comment by jasiek** [comment]
+1

**Comment by whicharuna** [comment]
+1

**Comment by DavidKutik** [comment]
+1

**Comment by OdinLin** [comment]
+1

**Comment by manokaran3529** [comment]
+1

**Comment by cCloudSky** [comment]
+1

**Comment by MutterCai** [comment]
+1

**Comment by susheel** [comment]
+1

**Comment by dementiacs** [comment]
+1

**Comment by mkjoerg** [comment]
+1

**Comment by helara** [comment]
+1

**Comment by scizeron** [comment]
+1

**Comment by deniseschannon** [comment]
Keycloak has been added to Rancher 2.0.

With the release of Rancher 2.0, development on v1.6 is only limited to critical bug fixes and security patches.

**Comment by cachaldora** [comment]
@deniseschannon do you have any plans to add OpenID Connect provider?

**Comment by sheng-liang** [comment]
We need to support OpenID Connect.

**Comment by cachaldora** [comment]
When will you include this request on your roadmap?
GitHub authentication uses oauth2 and you already have support for GitHub. It should be easy to create a custom oauth2 authentication module and it would be very nice 😃

**Comment by vincent99** [comment]
@cachaladora The actual authentication flow is a [small portion of the work](https://github.com/rancher/rancher/issues/1317#issuecomment-239398843) required to properly support a provider.  And everything sounds easy when you're telling someone else they should be doing it.

**Comment by cachaldora** [comment]
@vincent99 you already have support to oauth2 in your github provider, so I’m telling that it’s easy because you already have it. If you duplicate github provider and add some customization fields you will get a generic oauth2 provider. It’s not like build a new controller from scratch. Anyway, there is many people requesting this feature and there is people also request an alternative way of authentication to use with a reverse proxy. There are people requesting this features for at least 2 years, it would be good to understand if it will be in your roadmap.

**Comment by jasiek** [comment]
I am confident this would come into fruition if there was a bounty attached to this ticket. Perhaps this is the right way of making this happen?

I'll chip in £100. Who else wishes to contribute?

**Comment by erfangc** [comment]
Is this feature a call to become compatible with generic OAuth 2.0 IdP over OIDC or Google OAuth specifically? The former seems fitting for enterprises whereas the latter seems like it would just cause another issue to be raised down the line for a slew of other providers

Same for SAML 2.0 providers

**Comment by vincent99** [comment]
Support is being added for Google "G Suite" authentication in 2.3.  As I've said before:
  - The OAuth portion of authenticating a user is a small portion of what is required.  Rancher lets you configure authorization for clusters/projects/etc.  To manage that you need to be able to find and identify other people.  Getting info about users (other than yourself) and groups is not part of OAuth, (or OIDC, or SAML); it's custom code to talk to the unique API for each provider.
  - We provide commercial support for enterprise customers.  We test only the providers we support, and support only the providers we test.

**Comment by erfangc** [comment]
Thanks for that explanation. Can you help us understand why does rancher need to lookup other users when performing RBAC decisions on the authenticated principles?

If an IdP is configured for OIDC, then rancher should be able to extract the “sub” claim from the signed JWT token and map the “sub” to an internal user id right? “subs” that have not previously been seen can be given the default project and cluster role as needed, I agree that is definitely optional 

In terms of the ability to search the user database from an IdP that can probably never be properly supported generically as OIDC providers can themselves federate identity from a variety of other sources

**Comment by vincent99** [comment]
Yes logging in and figuring out who "you" are is easy, and what OIDC is for.  What it does not address is who other people are in that system, as you said.

One of the main things Rancher does is manage RBAC roles and bindings what resources "you" are supposed to be able to access.  To define those rules, someone else (an admin, project owner or cluster owner) needs to be able to look up you (or Steve in accounting, or the "development" group that they are not themselves a member of, or the new guy that just joined and hasn't even logged into Rancher yet, or whatever) and add role bindings for them.

Without that capability, just logging on its own is fairly useless to us.  To do anything with it a user will need to grant permissions to an unwieldy LDAP DN string, GUID, or random number that they may have no way to even find for another person.

For example, for GitHub we store integer user IDs since user names can be changed and reused.  But you probably have no idea what your own user number is, let alone how to find mine.  If GitHub were implemented as generic OAuth it would be totally unusable (or a security landmine when you configure it to use username since that sounds like a good idea at the time).

**Comment by erfangc** [comment]
Upon a new identity logging in the first time, rancher can store the standard JWT token claims: `sub`, `name`, `email`, `given_name`, `family_name` and make it available for the admin/project owner to search over, instead of having to query the original IdP. These claims are standard OIDC claims of the `id_token`. Rancher can  reject tokens that do not contain a `sub` and `email` (sub is unique per IdP). I understand this might be a departure or a bifurcation of the search and grant permission functionalities that's already been built, so not advocating for it as a replacement but rather as an additional option to consider

I imagine most customers and users would be content with the option of giving default roles and access to subjects through a connection that rancher has not seen before, then having an admin user search through the local set than not having OIDC connectivity at all

**Comment by vincent99** [comment]
That doesn't let you refer to users that haven't been into rancher yet, which is needed when using a public IdP full of users that shouldn't be able to. (The obvious is "just store the email and match it up later", but what's the canonical form of e.g. a Gmail address when dots are optional?  Every other email provider?)

Or have any mechanism for syncing (other than people logging in) or cleaning up (when they can't anymore).  Or handling eventual duplicates when steve@company.com leaves and that email is given to someone else. Or (probably) handle nested groups you're not directly a member of. Etc.

I'm sorry but I'm not going to keep doing this.  The bottom line is we don't like supporting half-baked solutions, and the demand from customers isn't there to force providing one.

**Comment by cachaldora** [comment]
@vincent99 I think that everybody already noticed that you don't want implementing this feature.

It would be very nice to use Rancher as many other software solutions that use its own local authentication database and uses a header or an environment variable or OAuth to know who is trying to access the system and if the user has authorization. This feature for me and for my company would be great because everything but Rancher is working like this for us.

Btw, I think you bringing some questions about how to manage the users that, in my opinion, should be a concern of who uses this feature and not to you.

**Comment by JohnCMcDonough** [comment]
@vincent99 I don't know if that's a good enough reason. We have tons of tools here, Gitlab being one of those, where you can't manage access for a user that hasn't logged in yet. That's not the end of the world, and it's better than not having OIDC at all. Ideally, we'd be able to use the same Identities that we use across all of our other software.

Regardless, in our case we would only have a single provider, that will always have unique `sub` claim. We have a finite number of predefined roles that we're planning on pushing from an external system into Rancher Server using a Service Account. We aren't allowing users to use Rancher to manage access. Only the ServiceAccount that pushes those roles will be able to change roles and bindings. The functionality of searching for users isn't necessary for us.

At the very least, you should consider adding some kind of user impersonation like what was described by @cachaldora above and in this PR (I can't find the original issue): https://github.com/rancher/rancher/pull/21396

**Comment by lukasmrtvy** [comment]
Maybe would be enough to implement support for https://github.com/dexidp/dex

**Comment by crazedmeph** [comment]
@vincent99 Im sorry but everything you said is not valid in any way. As @JohnCMcDonough already said since we have a lot of these SAML integrations we already do not manage users until they log in. And to be honest from a large organization standpoint we do not want to manage users inside of rancher. That one of the main reasons of using an IDP, allowing you to manage users in a central place. Rancher is but one of the applications people are using, allowing OpenID Connect will just give easier setup and support and allowing upstream redirects for more complicated business logic. I am not going to lie, why OpenID Connect is not implemented on something that is so far ahead of the curb is beyond me. We can still pull all of the information we need out of a OpenID connect setup just like SAML. So please for the love of all that is holy make our life easier and future proof rancher before it gets pushed out of the way by some other software, because i promise that is enough to make me change to something else all by itself.

**Comment by ClientIco** [comment]
+1 
@Vincent99 we rely on OIDC and have an idp as well as Authorization/Policy server that we believe can integrate with Rancher - Can you indicate or give us a better indication how we could in your words get this standardize so that Authentication/Authorization is sorted ? : 

> 

> - The OAuth portion of authenticating a user is a small portion of what is required. Rancher lets you configure authorization for clusters/projects/etc. To manage that you need to be able to find and identify other people. Getting info about users (other than yourself) and groups is not part of OAuth, (or OIDC, or SAML); it's custom code to talk to the unique API for each provider.
> 
> - We provide commercial support for enterprise customers. We test only the providers we support, and support only the providers we test.


Is the only way to become an enterprise customer ?

**Comment by skyscooby** [comment]
We really need this too.. Rancher in general seems to only support enabling one authentication method and the Active Directory integration only supports a single domain.

Gitlab (EE) supports many AD domains and auth providers.. and is also an oauth provider.. We want to use Gitlab's Oauth service to allow rancher accounts to come from a list of possible directories. We are fine with having to assign users to groups within rancher itself and only after they have logged in once. In this situation when a user is disabled in gitlab they shouldn't be able to log into rancher any longer so we don't really care if their account is automatically removed from rancher or not provided they can't access it.

**Comment by lindhe** [comment]
Just to clarify – is this issue about authentication in **(A)** RKE clusters, **(B)** any cluster that Rancher can manage or **(C)** just for the Rancher Web GUI?

**Comment by jasiek** [comment]
For me, it was the rancher web gui.

**Comment by skyscooby** [comment]
The authentication would need to extend to B as well.. at least via the proxy services rancher provides for access to a cluster... obviously it wouldn't be of value if we could not assign these users with cluster access level (eg the ability to generate a proxied kubecfg to a cluster they can access)

**Comment by thies226j** [comment]
It would be immensely helpful if an general option for oidc providers would be implemented.
Rancher is currently the only software in use that doesn't allow for this in our company.

**Comment by evilhamsterman** [comment]
It also baffles me that Rancher doesn't support generic OIDC or SAML providers. Every other service I've worked with may have some simplified options for tested providers that have a preconfigured setup where you only have to add a few bit of data like certificates. But they also have a custom option. Some rebuttals to @vincent99 

> That doesn't let you refer to users that haven't been into rancher yet, which is needed when using a public IdP full of users that shouldn't be able to. (The obvious is "just store the email and match it up later", but what's the canonical form of e.g. a Gmail address when dots are optional? Every other email provider?)
Like many others have said we are OK with not being able to find a user until they login. The canonical name issue is already solved and is part of the spec, regardless of what addresses and aliases the user has available (including dots and pluses in Gmail) the IdP will always send one canonical name. It would the defeat the purpose if it did otherwise because of the issue mentioned.
> 
> Or have any mechanism for syncing (other than people logging in) or cleaning up (when they can't anymore). Or handling eventual duplicates when [steve@company.com](mailto:steve@company.com) leaves and that email is given to someone else. Or (probably) handle nested groups you're not directly a member of. Etc.
Again let us worry about that. We can either use other tools to preprovision and deprovision them with the API, or we just don't worry about adding them till needed. We do that with some tools where as part of their onboarding they login to the various sites they need. We don't reuse emails for historical reasons and actually keep all old accounts around just deactivated so that we can reactivate them if they return to the company, or to maintain history in systems like version control. As long as they are deactivated they can't login and that's what we care about. As to syncing groups and memberships, the way things like Gitlab handle it it they also let you connect to a Directory like AD. That way can get group membership from AD and authentication from SAML. Or you can provide information in the SAML assertion that provides information like group membership to provision on the fly when they log in.
> 
> I'm sorry but I'm not going to keep doing this. The bottom line is we don't like supporting half-baked solutions, and the demand from customers isn't there to force providing one.
This doesn't need to be half-baked, there are plenty of services out there that support Generic OIDC and/or SAML. They are standards for a reason. Yes you may have to put provisos and limitations if it isn't a fully supported solution because every IdP adds their own extensions, but you should do the same and have a generic OIDC/SAML support that is used by every service and then add the extra features on top of that. These are all problems that have been solved by other solutions.

You may not see the demand because there are a lot of companies that will look at a product and just not consider it if can't support their authentication provider.

**Comment by jescarri** [comment]
We are starting to evaluate RKE and Rancher, we have been running kubernetes for more than 3 years, we run multitenant clusters and we have a good workflow to onboard developers to their namespaces, getting them a kubeconfig etc, the lack of support for general OIDC providers like dex might be a deal breaker for some organizations, as @evilhamsterman pointed out.

We use AD Groups + Dex + Rolebindings to Ldap Groups to authorize users, it works great, I hope soon rancher and RKE supports generic OIDC providers.

**Comment by lukasmrtvy** [comment]
Related https://github.com/rancher/rancher/issues/28804 ?

**Comment by wzrdtales** [comment]
https://github.com/coreos/go-oidc

**Comment by Blackclaws** [comment]
Definitely something that I see as a massive value add. I've been trying to get rancher to work with Keycloak SAML using the tutorial but its been going less than stellar. The OIDC workflow is just much more modern than SAML and is easier to setup for web applications anyway.

**Comment by skurfuerst** [comment]
Hey,

I've implemented a small proxy which bridges GitHub's and GitLab's APIs for usage with Rancher: https://sandstorm.de/de/blog/post/making-rancher-2-and-gitlab-oauth-authentication-work-together.html

All the best,
Sebastian

**Comment by cuchac** [comment]
Hello,

I can confirm proxy from @skurfuerst works like a charm! Great work, thank you! I've prepared Docker image for easy installation if someone is interested https://github.com/wpj-cz/rancher-gitlab-proxy

**Comment by merlindorin** [comment]
A lot of passionate debates, a lot of custom solutions (we also have one xD)... and finally a milestone!

🎁~~ enjoy ~~🎁

https://github.com/rancher/rancher/issues/30280

**Comment by bemanuel** [comment]
> Hey,
> 
> I've implemented a small proxy which bridges GitHub's and GitLab's APIs for usage with Rancher: https://sandstorm.de/de/blog/post/making-rancher-2-and-gitlab-oauth-authentication-work-together.html
> 
> All the best,
> Sebastian

For some reason for me isn't works.

**Comment by metal3d** [comment]
I don't understand why "OpenID Connect" is not implemented with universal configuration. e.g. we're using Apero CAS that has openidconnect. I connected Harbor and many others servers with "OIDC" configuration (we only need to give the well-known uri, client id, secret and some claims to get)

Why do you restrict configuration to a limited list ?

**Comment by wzrdtales** [comment]
b/c they have some artificial reasons to not do so.

to quote

> Without that capability, just logging on its own is fairly useless to us. To do anything with it a user will need to grant permissions to an unwieldy LDAP DN string, GUID, or random number that they may have no way to even find for another person

The reason was they just don't want to b/c the authenticating attribute might be something weird. that would be easily solved by just requiring either username or email as the identifier.

The  strange thing is I don't think anyone at rancher really took this conversation serious. Given, that they compare LDAP auth which is supported by rancher itself. Right now the only way to support openid is to use keycloak (which is verifying over saml which is pretty much the same as oidc just in bad). But honestly, refrain from keycloak only for doing this: keycloak is a truly horrible piece of software.

**Comment by Blackclaws** [comment]
> b/c they have some artificial reasons to not do so.
> 
> to quote
> 
> > Without that capability, just logging on its own is fairly useless to us. To do anything with it a user will need to grant permissions to an unwieldy LDAP DN string, GUID, or random number that they may have no way to even find for another person
> 
> The reason was they just don't want to b/c the authenticating attribute might be something weird and not immutable. that would be easily solved by just requiring either username or email as the identifier and letting the user know that this has the implication of loosing authentication with the system when the email would change for that user ( which does not happen in enterprise context anyways). However most system do have a unique id they deliver on top of whatever info rancher asks for. And why exclude all the systems that do deliver a real uid, just for the sake of avoiding the ones that don't.
> 
> The strange thing is I don't think anyone at rancher really took this conversation serious. Given, that they compare LDAP auth which is supported by rancher itself. Right now the only way to support openid is to use keycloak (which is verifying over saml which is pretty much the same as oidc just in bad). But honestly, refrain from keycloak only for doing this: keycloak is a truly horrible piece of software.

Its less artificial than you think. Yes you could say that email address or username is a requirement however sometimes that doesn't work. We have service-accounts that do not have emails associated with them for example. In general I think more flexibility in the domain of how users and groups are handled would help a lot. But since the fields where you add other people are always sourced and aren't freetype (not really) this is indeed an issue right now. You can't even add other ldap users to permissions with a local account under some circumstances. 

The drawback of not allowing a search on users or groups is definitely still there with oidc, however that is mitigated easily by just adding an ldap connection or as you suggested just allowing freeform entries for emails or the likes.

Keycloak itself isn't that bad once you get the hang of the idiosyncratic way they do things. There aren't really a lot of free oidc alternatives out there either.

**Comment by wzrdtales** [comment]
Be assured, I have good reasons to reject it and call it a horrible piece of software which it is, I had to work with it, I had to write extensions for it (even more horror) and had customers claiming it can do things which is why they decided to use it, and indeed it doesn't. my concerns with keycloak is misplaced in this issue, so I hope you forgive that I wont comment further since I don't wanna spam here. 

> Yes you could say that email address or username is a requirement however sometimes that doesn't work. 

As I also said, most OIDC do provide a unique identifier, with those rancher shouldn't have any problem at all.

**Comment by wzrdtales** [comment]
And to elaborate why I exaggerated with "artificial reasons". They're artificial b/c everyone else in the whole industry world wide has no problem with it, except rancher.

**Comment by Blackclaws** [comment]
You can't do a lookup for those identifiers however. That is the crux of the issue. If you have a system where you can't lookup the id of the user to add permissions to you need a separate system attached that provides this capability.

Most software I have seen that provides OIDC either has a separate connection to ldap as a user store (which might or might not have the oidc unique id, usually it should), or users only become discoverable after first login (which is a what a lot of the "register with oidc" types of software do).

Nextcloud is an example of a configurable former while Gitlab is firmly part of the second family. I think either approach has its merits, a best of both worlds scenario would probably be to provide either an ldap connection or create and store on first use oidc users. That way you have an additional constraint, which is that a user you want to add permissions for has to have logged in at least once. But that's all the drawback.

**Comment by wzrdtales** [comment]
The UID is only for long term auth, for the registration you will still use the secondary param. As soon as you connected your created user with the UID, the UID is perfectly fine serving and then you indeed **can** look them up.

**Comment by wzrdtales** [comment]
But here a solution that doesn't require **opinions**. Which is essentially the problem here, not the problem itself.

If rancher would make that part modular to easily add own authenticator modules and leave it to the community, we can just stop discussing here and a solution for this problem would arrive by tomorrow.

**Comment by kremers** [comment]
+1

**Comment by charlie-roosen** [comment]
+1
We are moving towards running everything on Kubernetes with OpenID Connect for authentication.  We would like this for the Rancher Server web interface.

I suggest they look at how Grafana is supporting Oauth2

https://grafana.com/docs/grafana/latest/auth/generic-oauth/

In particular, using JMESPath to map users to roles.

**Comment by metal3d** [comment]
OK so that means that for now, we cannot connect users to our OIDC for a long time?

It's a pity... because we had decided to use Rancher and potentially try a partnership, but if only GitHub and KeyCloak are supported, we can't authenticate our users (we don't want the LDAP server to be possibly impacted by Rancher, hence the need for OIDC, CAS, or other protocol barriers)

I understand that you have opinions (positive or negative) about OIDC, but it is a widely used and mastered protocol by professionals. All our applications use at least CAS or OIDC, Rancher is an exception which unfortunately will block its use in our company :(

One thing is for sure, I should have checked earlier... no malice on my part, I really thought that OIDC was already supported by Rancher (until now I used internal authentication). 

I'm not trying to accuse you of doing things wrong, but I'll let you imagine my disappointment because I love Rancher and I really wanted it to be part of our internal product panel.

(Excuse my English if you think that I'm aggressive, I'm not, I really love Rancher)

**Comment by evilhamsterman** [comment]
> You can't do a lookup for those identifiers however. That is the crux of the issue. If you have a system where you can't lookup the id of the user to add permissions to you need a separate system attached that provides this capability.
> 
> Most software I have seen that provides OIDC either has a separate connection to ldap as a user store (which might or might not have the oidc unique id, usually it should), or users only become discoverable after first login (which is a what a lot of the "register with oidc" types of software do).
> 
> Nextcloud is an example of a configurable former while Gitlab is firmly part of the second family. I think either approach has its merits, a best of both worlds scenario would probably be to provide either an ldap connection or create and store on first use oidc users. That way you have an additional constraint, which is that a user you want to add permissions for has to have logged in at least once. But that's all the drawback.

I think this shows that Rancher has a fundamental misunderstanding of what OIDC (and SAML) is for and that is leading to their refusal to implement it.

OIDC is for Authentication that's it, it is used for verifying someone is who they say they are. Authorization, i.e. looking up groups and assigning permissions, is left to the software that the IDP is sending the authentication data to, i.e. Rancher/Gitlab/Github/Nextcloud/etc. Like has been said many times that could be a local db, a local db that utilizes metadata sent with the authentication packet a connection to an LDAP instance, an API call to another service, or something else. 

But it comes down to OIDC/SAML are a secure method for offloading authentication to a third party no more no less. Once they understand that then it means they have to fundamentally shift how they are handling Authentication and Authorization and doing that is hard. But they've apparently decided they'd rather lose sales and revenue than the hard right thing and fixing it.

**Comment by jescarri** [comment]
We are on the same boat, and we do love rancher, we where able to gain some time by using yet another proxy + impersonation + dex, it works great but we would love to see support for external OIDC providers, IMO Dex is as simple as it gets, I haven't completely dived into the topic but from my experience there's nothing that impeedes having the classic rancher auth/athorization + an external OIDC provider to coexist.

I hope rancher engineering listens to this long asked feature.

**Comment by letian0805** [comment]
Is there any plan to support CAS?

**Comment by arash-bizcover** [comment]
More than 5 years from this issue, yet they haven't understood how important this feature is. 
![image](https://user-images.githubusercontent.com/62287461/147621855-f3d13a61-4a6b-4964-8a0a-735276535f45.png)

**Comment by nobiit** [comment]
How is this problem? Does anyone have any other solution?

**Comment by bradjones1** [comment]
@arash-bizcover @nobiit Pinging (or worse, posting passive-aggressive memes) in the issue queue is no way to get features out of a product most people use for free.

**Comment by arash-bizcover** [comment]
> @arash-bizcover @nobiit Pinging (or worse, posting passive-aggressive memes) in the issue queue is no way to get features out of a product most people use for free.

haha, envy your patience in not unsubscribing from this issue. Relax, let go, "Enjoy your life"

**Comment by flyte** [comment]
> the demand from customers isn't there to force providing one

> a product most people use for free

I doubt any of those paying customers started paying without first using Rancher free of charge.

> we don't like supporting half-baked solutions

I just hooked up ArgoCD to my client's GitLab OIDC provider without any fuss. They bundle Dex along with their product in order to offload this type of issue and get a lot of functionality with minimal support requirement.

Maybe if Rancher just included Dex support, then we'd all go away and leave them alone.

Maybe some more of us might even _pay for the software_.

**Comment by evilhamsterman** [comment]
Yeah they seemed to have just decided to bury their heads in the sand and ignore the fact that this is a solved problem by hundreds of other apps including their direct competitors

For example
* OpenShift OIDC https://docs.openshift.com/container-platform/4.7/authentication/identity_providers/configuring-oidc-identity-provider.html
* Tanzu OIDC https://docs.vmware.com/en/VMware-Tanzu-Kubernetes-Grid/1.3/vmware-tanzu-kubernetes-grid-13/GUID-mgmt-clusters-enabling-id-mgmt.html

**Comment by Blackclaws** [comment]
Yeah, the SAML auth is a pain to setup and OIDC is a widely available and deployed solution. OIDC also supports group claims etc. so I see no reason why it shouldn't be possible to just have a generic OIDC provider setup.

**Comment by evilhamsterman** [comment]
@Blackclaws I personally find SAML a bit easier to setup than OIDC, but both support similar features, and OIDC is generally regarded as the more secure option. The matter still stands this is a solved problem Rancher just doesn't think it is important enough to dedicate resources to it. Maybe SUSE will realize it eventually

**Comment by nynymike** [comment]
There is no discussion any more about whether OpenID or SAML is more secure.  Fundamentally, XML is too hard to parse, and XML canonicalization is easy to get wrong.  There have been so many SAML library vulnerabilities, this is not even an area of debate anymore. And as SAML doesn't support mobile, it is essentially frozen in time (never a good thing for a standard to adapt to new attacks).

Also, talking about what OpenID Provider to use--Keycloak, Gluu Server, Ory, Dex or any of the [other](https://openid.net/certification/) excellent implementations listed in the certifications page--is a *non-sequitur*. 

As a Relying Party (RP), the important conversation that needs to happen is what features of OpenID Connect should the Rancher UI support. 

The goal of OpenID was to make simple things simple, but to make complex things possible. You can do a very minimal implementation, like implicit flow, even though there is are [contrary to current security recommendations](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-19.txt#section-2.1.2). 

However, if you implement enough of the OpenID features, you can also make it very secure. Even secure enough for banking--this is why there is a [profile](https://openid.net/wg/fapi/) which uses every security feature known to man. 

But security comes with friction. Not only does the server have to support those features, but the developer writing the Rancher plugin will have to build the corresponding client features. And... admins who connect Rancher to their domain OpenID Provider will have to do things, like generate keys and make valid JWKS files.

So the design that needs to happen is to figure out which features [OpenID](https://openid.net/connect) and [OAuth](https://tools.ietf.org/wg/oauth/) features to use. And how can the Rancher Admin UI reduce the technical burden, and offer a range of security options that allow operational admins to self-select security features based on risk.

**Comment by evilhamsterman** [comment]
I think getting into deep discussions of SAML vs OIDC is distracting from the goal of this issue. I should have been clearer, from an administrative side SAML is easier to implement since you just need to paste some URLs and a cert between the IdP and Sp, vs going through an Oauth workflow to setup the connection. 

The point of this issue is that Rancher should support generic IdPs, preferably with OIDC but I'd take SAML at this point too. The problem is they apparently equivocate authorization, the goal of OIDC/SAML, with authorization which is separate and should be handled separately. They claim this is an unsolvable problem, see the early comments, despite the evidence to the contrary of many other projects solving this problem including their direct competitors.

**Comment by belgaied2** [comment]
Can we please implement this feature? There is a lot of interest from the community and also paying Rancher customers.

**Comment by bramamad** [comment]
We would like the OIDC login, too. We normally use Azure AD as our OIDC provider and it works wonderful with other opensource applications, eg. "harbor" -> https://goharbor.io/docs/2.6.0/administration/configure-authentication/oidc-auth/

**Comment by nynymike** [comment]
The current instructions are KeyCloak specific. As OpenID Connect is an open standard, this is not what is needed. Any conformant OpenID Provider should be supported.

**Comment by AnthonySLWhite** [comment]
Is it possible today to integrate a service like ORY Hydra for authentication in Rancher? 
Or is it only limited to the custom integrations?

**Comment by nynymike** [comment]
ORY Hydra, Janssen Project, Gluu Server (Google, Okta, MS Azure AD), Ping Identity, Forgerock, WS02... all modern IAM platforms... support [OpenID Connect](https://openid.net/connect).  As long as the Rancher UI supports OpenID, it can support all open source, commercial, and SaaS identity providers. Seems like a no-brainer? It should be...

**Comment by evilhamsterman** [comment]
I've given up and we've decided to pass on Rancher, we used OneLogin which isn't supported so there was no way to integrate with our systems. We've since switched to Okta which they do support but we've already moved on and found other solutions. So they mark down at least one missed sale due to this issue.

**Comment by AndreKR** [comment]
@evilhamsterman
> but we've already moved on and found other solutions

With what did you end up?

**Comment by evilhamsterman** [comment]
@AndreKR We ended up just managing everything with GitOps using Fluxcd. Monitoring with our existing Zabbix, and if we need a dashboard the vanilla Kubernetes dashboard

**Comment by dimq** [comment]
Hey, still looking into that to connect it to gitlab, I just see this look like a generic oidc config :
https://github.com/rancher/rancher/blob/master/pkg/auth/providers/oidc/oidc_provider.go
I tried to set up it through yaml, but I don't know how to activate it. More on it, am I right to suppose it's the generic oidc implementation ?

**Comment by mriesen** [comment]
Same here... in my case it's GitLab which cannot be used as an IdP for Rancher.
Is there any plan to support OIDC in general?
It'd be great to have the option to integrate any OIDC provider instead of the specific ones.
I think that'd be the way to go, to support a standard instead of specific providers.

**Comment by nicewrld** [comment]
Still running into this, being forced to use SAML in $currentday

**Comment by nynymike** [comment]
It's really dissapointing that we have to put up with XML security issues today when there is a more secure, convenient alternative.

**Comment by txj-xyz** [comment]
Following this for implementation

**Comment by gaktive** [comment]
Depending on how development goes here, if UI could get test builds once they're in a ready state, work can be done in parallel prior to integration. Our related ticket: https://github.com/rancher/dashboard/issues/9820

**Comment by stefannilsson** [comment]
+1 here

**Comment by crobby** [comment]
@joesims22 
QA Testing Considerations:
1. Engineering manual testing has mostly been done with Keycloak and Okta-dev.  It is a good idea to try out some additional providers.
2. Be sure that various global/cluster/project role bindings are working as expected with generic-oidc users.  (keep in mind that the "search" functionality is faked in the typical Rancher way since OIDC is not searchable via a service account or anything like that)
3. Be sure  that tokens created by generic-oidc users are working as expected.
4. Make sure that generic-oidc users (and their bindings) are cleaned-up when deactivating generic-oidc.  There was no new code added for this but it should be "standard" authConfig behavior.

Regression Considerations:
1.  It is probably worth testing keycloak-oidc in addition to generic-oidc.  There is some common code between the two.  Great care was taken to not change anything "material" that would affect keycloak-oidc, but it's definitely worth checking.

**Comment by joesims22** [comment]
Validated on v2.9-head `3c4ccdc`. 

The following test cases were ran with GenericOIDC using both Keycloak and Okta auth providers:


Test case | Status
-- | --
Verify P0 auth checks via Generic OIDC using | PASS
Verify group membership via Generic OIDC | PASS
Verify nested group membership via  Generic OIDC | PASS
Enable, disable, and re-enable auth GenericOIDC auth provider | PASS
Verify GenericOIDC is shown as enabled in auth config | PASS
 Disable Generic OIDC auth provider via kubectl and verify all resources are deleted | PASS
 Verify all related resources are deleted upon deactivating Generic OIDC auth provider | PASS
 Auth Upgrade check for Generic OIDC | PASS
 Verify P0 auth checks on regular Keycloak OIDC auth provider | PASS
 Verify all related resources are deleted upon deactivating on Keycloak auth provider | PASS
